guvnor-asset-mgmt/guvnor-asset-mgmt-backend/src/main/java/org/guvnor/asset/management/backend/service/RepositoryStructureModelLoader.java

```diff
@@ -0,0 +1,155 @@
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import javax.inject.Inject;
+
+import org.guvnor.asset.management.model.RepositoryStructureModel;
+import org.guvnor.common.services.project.model.POM;
+import org.guvnor.common.services.project.model.Project;
+import org.guvnor.common.services.project.service.POMService;
+import org.guvnor.common.services.project.service.ProjectService;
+import org.guvnor.common.services.shared.metadata.MetadataService;
+import org.guvnor.structure.repositories.Repository;
+import org.uberfire.backend.server.util.Paths;
+import org.uberfire.backend.vfs.Path;
+import org.uberfire.io.IOService;
+
+import static org.guvnor.structure.repositories.EnvironmentParameters.*;
+
+public class RepositoryStructureModelLoader {
+
+    @Inject
+    private POMService pomService;
+
+    @Inject
+    private ProjectService<? extends Project> projectService;
+
+    @Inject
+    private ManagedStatusUpdater managedStatusUpdater;
+
+    @Inject
+    private MetadataService metadataService;
+
+    public RepositoryStructureModel load( final Repository repository,
+                                          final String branch,
+                                          final boolean includeModules ) {
+        if ( repository == null ) {
+            return null;
+        }
+
+        final Project project = projectService.resolveToParentProject( repository.getBranchRoot( branch ) );
+
+        if ( project != null ) {
+            return getModel( repository,
+                             includeModules,
+                             project );
+
+        } else {
+            return getModel( repository,
+                             branch );
+        }
+    }
+
+    private RepositoryStructureModel getModel( final Repository repository,
+                                               final String branch ) {
+
+        final RepositoryStructureModel model = new RepositoryStructureModel();
+
+        final Boolean managedStatus = getManagedStatus( repository );
+
+        //if no parent pom.xml present we must check if there are orphan projects for this repository.
+        final Set<Project> repositoryProjects = projectService.getProjects( repository,
+                                                                            branch );
+        if ( repositoryProjects.isEmpty() && managedStatus == null ) {
+            //there are no projects and the managed attribute is not set, means the repository was never initialized.
+            return null;
+        }
+
+        if ( managedStatus != null ) {
+            model.setManaged( managedStatus );
+        }
+
+        model.setOrphanProjects( new ArrayList<>( repositoryProjects ) );
+        for ( Project orphanProject : repositoryProjects ) {
+            final POM pom = pomService.load( orphanProject.getPomXMLPath() );
+            model.getOrphanProjectsPOM().put( orphanProject.getSignatureId(),
```

Comment:
For all I can think of, is that this signature id works even when the GAV is incomplete or not yet set. The signature id is the project class type, in our case always KieProject + the path to project root. Feels to me like the project root would be good enough.